nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,172 @@
+# or logic issues as a packet on a gateway is a sure way to know that it really arrived, however
+# it poses a difficulty when debugging and also filtering. All traffic on gateways will simply
+# be UDP WH encrypted packets either towards DERP or towards the peer itself, meaning testing
+# batching partially is impossible without decrypting WG traffic(in order to check if one part of libtelio
+# is batchet or not). Thus capturing on clients is preferred at the moment.
+CLIENTS = [
+    pytest.param(
+        _generate_setup_parameter_pair([
+            (ConnectionTag.DOCKER_CONE_CLIENT_1, ["stun"]),
+            (ConnectionTag.DOCKER_CONE_CLIENT_2, ["stun"]),
+            (ConnectionTag.DOCKER_CONE_CLIENT_3, ["stun"]),
+        ])
+    ),
+]
+
+
+class Gateway:
+    def __init__(self, name: str, tag: ConnectionTag):
+        self._name = name
+        self._tag = tag
+        self._conn = None
+
+    def __repr__(self):
+        return f"Gateway({self._name}, {self._tag}, {self._conn})"
+
+
+@pytest.mark.asyncio
+async def test_histogram():
+    bo = Histogram(10)
+    for _ in range(10):
+        bo.add_value(2)
+        bo.add_value(3)
+
+    for _ in range(50):
+        bo.add_value(4)
+
+    bo.add_value(9)
+
+    assert bo.get() == [0, 0, 10, 10, 50, 0, 0, 0, 0, 1]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("setup_params", CLIENTS)
+async def test_direct_batching(
+    setup_params: List[SetupParameters]
+) -> None:
+    async with AsyncExitStack() as exit_stack:
+        capture_nodes = [
+            Gateway(
+                "DOCKER_CONE_CLIENT_1",
+                ConnectionTag.DOCKER_CONE_CLIENT_1,
+            ),
+            Gateway(
+                "DOCKER_CONE_CLIENT_2",
+                ConnectionTag.DOCKER_CONE_CLIENT_2,
+            ),
+            Gateway(
+                "DOCKER_CONE_CLIENT_3",
+                ConnectionTag.DOCKER_CONE_CLIENT_3,
+            ),
+        ]
+
+        async def run_batch_observer(gw: Gateway, mesh_ip: str):
+            print(f"Running batch observer for gateway: {gw._name} on mesh ip: {mesh_ip} ")
+            
+            await asyncio.sleep(5)
+            print(f"Running batch observer for gateway after delay: {gw._name}")
+            async with BatchObserver(f"{gw._name}", gw._conn).run() as bo:
+                await asyncio.sleep(15)
+
+                hs_inc = bo.get_histogram(
+                    20, 200, ObservationTarget(Direction.Incoming, mesh_ip)
+                )
+                hs_out = bo.get_histogram(
+                    20, 200, ObservationTarget(Direction.Outgoing, mesh_ip)
+                )
+                
+                print(">>>>>>> batching results vvv: ", gw._name)
+                print("hs_inc:", hs_inc)
+                print("hs_out", hs_out)
+                print(">>>>>>> batching results ^^^")
+
+        env = await exit_stack.enter_async_context(
+            setup_environment(exit_stack, setup_params)
```

Comment:
Better to misalign here

Reply:
Misalignment now is done on initialization by creating 2 meshnet peers and then adding another one after sleeping for a small amount of time.